pkg/typesystem/typesystem.go

```diff
@@ -667,167 +667,214 @@ func (t *TypeSystem) relationInvolvesExclusion(objectType, relation string, visi
 	return false, nil
 }
 
-// hasEntrypoints recursively walks the rewrite definition for the given relation to determine if there is at least
-// one path in the rewrite rule that could relate to at least one concrete object type. If there is no such path that
-// could lead to at least one relationship with some object type, then false is returned along with an error indicating
-// no entrypoints were found. If at least one relationship with a specific object type is found while walking the rewrite,
-// then true is returned along with a nil error.
-func hasEntrypoints(
+type RelationDetails struct {
+
+	// hasEntrypoints is true when a particular object+relation relates to at least one assignable object type.
+	// A relation such as `define document#viewer: [document#viewer]` implies that `hasEntrypoints=false`.
+	// A relation such as `define document#viewer: [user, document#viewer]` implies that `hasEntrypoints=true`.
+	hasEntrypoints bool
+
+	// hasCycle is true when a particular object+relation introduces a cycle in the model.
+	// A relation such as `define document#viewer: [document#viewer]` does NOT imply a cycle.
+	// hasEntrypoints=true implies that hasCycle=false.
+	// If hasEntrypoints=false, nothing can be said about hasCycle.
+	hasCycle bool
```

Comment:
I still don't know the difference between this and https://github.com/openfga/openfga/blob/b4f33fd3d7ec140e5ed4b3f94745656f5f930e03/pkg/typesystem/typesystem.go#L1294

That function returns cycles that `GetRelationDetails` does not. 🤷‍♀️ 

Ideally we only keep one of the two...